### PR TITLE
performance: use OPTIMISTIC transaction as the default

### LINKED ISSUE #751: use OPTIMISTIC mode default
## Feature Request

**Is your feature request related to a problem? Please describe:**

Starting from v3.0.8, newly created TiDB clusters use the pessimistic transaction model by default
but for dm we can set session level optimistic transaction as default to improve performance. 

ref https://pingcap.com/docs/dev/pessimistic-transaction/
**Describe the feature you'd like:**

Add optimistic txn session config connect to TiDB

**Describe alternatives you've considered:**

And merge multiple statements into one statement with multiple values in one transaction would help to improve performance.
**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
close #751 

### What is changed and how it works?
set default transaction mode to optimistic

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
    verify with different session cfg and run `show variables`

### CODE REVIEW

File: dm/config/task.go [hunk 1]
```diff
@@ -707,3 +715,14 @@ func checkDuplicateString(ruleNames []string) []string {
 	}
 	return dupeArray
 }
+
+func adjustTargetDB(dbConfig *DBConfig) {
+	if dbConfig.Session == nil {
+		dbConfig.Session = make(map[string]string, len(defaultSessionCfg))
+	}
+	for k, v := range defaultSessionCfg {
+		if _, ok := dbConfig.Session[k]; !ok {
```
Comment (lance6716) [inline dm/config/task.go:724]: seems variables are ignore-case, so we might use "to lower"

**Review by lance6716 — COMMENTED** [review]
rest LGTM

**Comment by lance6716** [comment]
LGTM

File: dm/config/task.go [hunk 2]
```diff
@@ -66,6 +68,11 @@ var (
 	defaultBatch                   = 100
 	defaultQueueSize               = 1024 // do not give too large default value to avoid OOM
 	defaultCheckpointFlushInterval = 30   // in seconds
+
+	// TargetDBConfig
+	defaultSessionCfg = map[string]string{
+		tidbTxnMode: tidbTxnOptimistic,
```
Comment (csuzhangxc) [inline dm/config/task.go:74]: Does a lower version of TiDB without optimistic works well (or without interrupt)?

**Review by csuzhangxc — COMMENTED** [review]
rest LGTM

File: dm/config/task_test.go [hunk 3]
```diff
@@ -645,3 +649,37 @@ func (t *testConfig) TestMySQLInstance(c *C) {
...
+		result   DBConfig
+		version  utils.TiDBVersion
+	}{
+		{
+			DBConfig{},
+			DBConfig{Session: map[string]string{}},
+			utils.TiDBVersion{0, 0, 0},
+		},
+		{
+			DBConfig{Session: map[string]string{"ANSI_QUOTES": ""}},
```
Comment (csuzhangxc) [inline dm/config/task_test.go:665]: key `SQL_MODE`, value `ANSI_QUOTES`?

File: dm/master/server_test.go [hunk 4]
```diff
@@ -503,12 +552,13 @@ func (t *testMaster) TestStartTaskWithRemoveMeta(c *check.C) {
 	c.Assert(server.pessimist.Start(ctx, etcdTestCli), check.IsNil)
 	c.Assert(server.optimist.Start(ctx, etcdTestCli), check.IsNil)
 
-	db, mock, err := sqlmock.New()
-	c.Assert(err, check.IsNil)
-	conn.DefaultDBProvider = &mockDBProvider{db: db}
+	verMock := t.initVersionDB(c)
```
Comment (lance6716) [inline dm/master/server_test.go:555]: defer and reset here?

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: dm/master/server_test.go (hunk 4 above)
Comment (GMHDBJD) [reply dm/master/server_test.go:555]: L556 do it

**Comment by lance6716** [comment]
LGTM

File: dm/master/server.go [hunk 5]
```diff
@@ -1027,17 +1027,47 @@ func parseAndAdjustSourceConfig(contents []string) ([]*config.SourceConfig, erro
...
 			return cfgs, err
 		}
 
+		fromDB.Close()
 		cfgs[i] = cfg
 	}
 	return cfgs, nil
 }
 
+func adjustTargetDB(ctx context.Context, dbConfig *config.DBConfig) error {
```
Comment (csuzhangxc) [inline dm/master/server.go:1044]: How about putting this function into `task.go` as a member of `TaskConfig`? we may need to use it in other places later.

File: pkg/utils/db.go [hunk 6]
```diff
@@ -424,3 +427,44 @@ func GetGTID(db *sql.DB) (string, error) {
 	val, err := GetGlobalVariable(db, "GTID_MODE")
 	return val, err
 }
+
+// ToTiDBVersion convert string to TiDBVersion
+// version format: 5.7.25-TiDB-v3.0.7
+func ToTiDBVersion(v string) (TiDBVersion, error) {
```
Comment (csuzhangxc) [inline pkg/utils/db.go:433]: TiDB-Lightning also has its implementation, https://github.com/pingcap/tidb-lightning/blob/e4de23b08259514ca20797696b6cd021a6a6de1a/lightning/backend/importer.go#L256

you may re-check with it.

**Review by csuzhangxc — COMMENTED** [review]
rest LGTM

File: dm/master/server.go (hunk 5 above)
Comment (GMHDBJD) [reply dm/master/server.go:1044]: circle import if move it to task.go because we import config in pkg.conn. 🤔

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**